browser/ai_chat/associated_link_content_unittest.cc

```diff
@@ -0,0 +1,270 @@
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/content/browser/associated_link_content.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/run_loop.h"
+#include "base/test/gmock_callback_support.h"
+#include "base/test/task_environment.h"
+#include "base/test/test_future.h"
+#include "brave/components/ai_chat/content/browser/page_content_fetcher.h"
+#include "brave/components/ai_chat/core/browser/associated_content_delegate.h"
+#include "chrome/test/base/chrome_render_view_host_test_harness.h"
+#include "chrome/test/base/testing_profile.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/test/mock_navigation_handle.h"
+#include "content/public/test/test_renderer_host.h"
+#include "content/public/test/web_contents_tester.h"
+#include "testing/gtest/include/gtest/gtest.h"
+#include "url/gurl.h"
+
+using base::test::RunOnceCallback;
+using testing::_;
+using testing::Return;
+using testing::StrictMock;
+
+namespace ai_chat {
+
+namespace {
+
+class MockPageContentFetcher : public PageContentFetcher {
+ public:
+  explicit MockPageContentFetcher(content::WebContents* web_contents)
+      : PageContentFetcher(web_contents) {}
+
+  MOCK_METHOD(void,
+              FetchPageContent,
+              (std::string_view, FetchPageContentCallback));
+};
+
+}  // namespace
+
+// Test wrapper that exposes private methods for testing
+class TestableAssociatedLinkContent : public AssociatedLinkContent {
+ public:
+  TestableAssociatedLinkContent(GURL url,
+                                std::u16string title,
+                                content::BrowserContext* browser_context)
+      : AssociatedLinkContent(std::move(url),
+                              std::move(title),
+                              browser_context) {
+    auto fetcher =
+        std::make_unique<MockPageContentFetcher>(web_contents_.get());
+    mock_fetcher_ = fetcher.get();
+    content_fetcher_ = std::move(fetcher);
+  }
+
+  ~TestableAssociatedLinkContent() override { mock_fetcher_ = nullptr; }
+
+  using AssociatedContentDelegate::set_cached_page_content;
+  using AssociatedLinkContent::DidFinishNavigation;
+  using AssociatedLinkContent::DocumentOnLoadCompletedInPrimaryMainFrame;
+  using AssociatedLinkContent::OnContentExtractionComplete;
+  using AssociatedLinkContent::OnTimeout;
+  using AssociatedLinkContent::web_contents_;
+
+  base::OneShotEvent* content_loaded_event() {
+    return content_loaded_event_.get();
+  }
+
+  void set_page_content_result(std::string content, bool is_video) {
+    EXPECT_CALL(*mock_fetcher_, FetchPageContent(_, _))
+        .WillOnce(RunOnceCallback<1>(content, is_video, ""));
+  }
+
+ private:
+  raw_ptr<MockPageContentFetcher> mock_fetcher_;
+};
+
+class AssociatedLinkContentTest : public ChromeRenderViewHostTestHarness {
```

Comment:
Adding new unit test in `//brave/browser` to test something that is in `//brave/components is forbidden`. cc @bridiver 
We should use `content::RenderViewHostTestHarness` and move this test into `brave/components/ai_chat/content/browser`.

Reply:
Fixed this just before I saw the comment :laughing: